---
title: "GenePop - general stats"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: inline
---

# Draft for the section general stats

## Parameters 

Inherited parameter and data from the import tab. 

```{r Parameters, tidy = TRUE}
filtered_data <-
  read.csv("../data/data-2023-09-11 (2).csv", header = TRUE)

selected_stats <-
  c("Ho", "Hs", "Ht", "Fis (W&C)", "Fst (W&C)", "Fis (Nei)", "Fst (Nei)")

n_rep = 10000
n_marker = 6
n_pop = 8

sequence_length <- length(6:11) 

num_cores <- 16
```

## data filtering convertion

```{r data filtering, tidy = TRUE, results='hide', message=FALSE}
library(hierfstat)
library(adegenet)
filtered_data <-
  data.frame(indv = paste(
    substr(filtered_data$Population, 1, 3),
    row.names(filtered_data),
    sep = "."
  ), filtered_data)
# Create mydata_genind
population <- filtered_data$Population
mydata_genind <- df2genind(
  X = filtered_data[, 6:11],
  sep = "/",
  ncode = 6,
  ind.names = filtered_data$indv,
  pop = filtered_data$Population,
  NA.char = "0/0",
  ploidy = 2,
  type = "codom",
  strata = NULL,
  hierarchy = NULL
)
mydata_hierfstat <- genind2hierfstat(mydata_genind)
```

## Missing data

```{r Missing,tidy = TRUE}
# Libraries
library(poppr)
library(heatmaply)
library(tidyr)
library(tibble)

missing_data <- info_table(mydata_genind, plot = FALSE, percent = TRUE, df = TRUE)
missing_data <- missing_data %>% spread(key = Locus, value = Missing  ) 
missing_data <- missing_data %>% column_to_rownames(var = "Population")

# heatmap
p <- heatmaply(missing_data, 
               dendrogram = "none",
               xlab = "", ylab = "", 
               main = "",
               scale = "none",
               margins = c(60,100,40,20),
               grid_color = "white",
               grid_width = 0.00001,
               titleX = FALSE,
               hide_colorbar = FALSE,
               branches_lwd = 0.1,
               label_names = c("Population", "Marker", "Value"),
               fontsize_row = 8, fontsize_col = 8,
               labCol = colnames(missing_data),
               labRow = rownames(missing_data),
               heatmap_layers = theme(axis.line=element_blank())
)
p
```

## Run basic.stats and render the result

```{r basic.stats,tidy = TRUE}
library(pegas)
library(dplyr)
library(tibble)

result <- basic.stats(mydata_hierfstat)  # hierfstat
df_resutl_basic <- as.data.frame(result$perloc)

# Weir and Cockrham estimates of F-statistics - FIS and FST
result_f_stats <- Fst(as.loci(mydata_genind), na.alleles = "")  # pegas
result_f_stats <- result_f_stats[, 2:3]
colnames(result_f_stats) <- c("Fst (W&C)", "Fis (W&C)")
result_f_stats <- merge(result_f_stats, df_resutl_basic, by = "row.names", all.x = TRUE)
colnames(result_f_stats)[10] <- "Fst (Nei)"
colnames(result_f_stats)[12] <- "Fis (Nei"
result_f_stats <- result_f_stats %>% column_to_rownames(., var = 'Row.names')
result_f_stats_selec <- result_f_stats %>% select(all_of(selected_stats))
result_f_stats_selec

```

## missing data / FIS

```{r missing, tidy = TRUE}
library(hrbrthemes)
library(tidyverse)

# Select Fis (W&C) from result_f_stats_selec
a <- result_f_stats_selec %>% 
  select(`Fis (W&C)`)

# Process the missing data
b <- as_tibble(missing_data) %>% 
  filter(row_number()==n()) %>%  
  rownames_to_column %>% 
  gather(variable, value, -rowname) %>%  
  spread(rowname, value) %>%  
  slice(1:(n() - 1)) %>% 
  column_to_rownames('variable')
colnames(b) <- ('Missing %')


# Merge the data frames
c <- merge(a, b, by="row.names",all.x=TRUE) %>% 
  column_to_rownames('Row.names')


# Plot with linear trend
p1 <- ggplot(c, aes(x = `Missing %`, y = `Fis (W&C)`)) +
  geom_point() +
  theme_ipsum()
p1

```



## G-statistic

```{r G-statistic, tidy = TRUE, cache=TRUE}
library(ggplot2)

# compute the Gstats
result_f_stats <- result_f_stats %>% mutate(GST = 1-Hs/Ht)
result_f_stats <- result_f_stats %>% mutate("GST''" = (n_pop*(Ht-Hs))/((n_pop*Hs-Ht)*(1-Hs)))
result_f_stats

# Plot with linear trend
p2 <- ggplot(result_f_stats, aes(x=GST, y=Hs)) +
  geom_point() +
  geom_smooth(method=lm , color="red", se=FALSE) +
  theme_ipsum()
p2

library(ggplot2)

# Compute the Gstats
result_f_stats <- result_f_stats %>% 
  mutate(GST = 1 - Hs / Ht,
         `GST''` = (n_pop * (Ht - Hs)) / ((n_pop * Hs - Ht) * (1 - Hs)))

# Plot with linear trend
p2 <- ggplot(result_f_stats, aes(x = GST, y = Hs)) +
  geom_point() +
  geom_smooth(method = "lm", color = "red", se = FALSE) +
  theme_ipsum()

p2

```

## Linkage Disequilibrium

### with poppr
The index of association was originally developed by A.H.D. Brown analyzing population structure of wild barley (Brown, 1980).

**Ia**: The index of association.
**p.Ia**: The p-value resulting from a one-sided permutation test based on the number of samples indicated in the original call.
**rbard**: The standardized index of association.
**p.rd**: The p-value resulting from a one-sided permutation test based on the number of samples indicated in the original call.

**pair.ia** calculates the index of association in a pairwise manner among all loci.

method = 1: Permute Alleles This will redistribute all alleles in the sample throughout the locus. Missing data is fixed in place. This maintains allelic structure, but heterozygosity is variable.

method = 4: Multilocus Style Permutation This will shuffle the genotypes at each locus, maintaining the heterozygosity and allelic structure.

```{r Linkage_poppr, cache=TRUE}
library(tictoc)
tic(msg = "pair.ia")
loci_pair <- pair.ia(mydata_genind, sample = n_rep, quiet = FALSE, method = 4,plot = FALSE)
loci_pair
toc()
```

## Panmixia - HW

```{r shuffle, tidy = TRUE, cache=TRUE}
library(poppr)
library(pegas)
library(dplyr)
library(tibble)
#Permute Alleles This will redistribute all alleles in the sample throughout the locus. Missing data is fixed in place. #This maintains allelic structure, but heterozygosity is variable.
tic(msg = "pair.ia")


es <- replicate(n_rep, (shufflepop(mydata_genind, method=1)))
fis_df <- numeric(sequence_length)

#for (i in 1:n_rep){
#  # Calculate the statistics for the i-th matrix
#  result_fis_stats <- as_tibble(Fst(as.loci(es[[i]]))) %>% select("Fis" )
#  # Assign values to the data frames
#  fis_df <- cbind(fis_df, result_fis_stats$Fis)
#}

library(foreach)
library(doParallel)

cl <- makeCluster(num_cores)
registerDoParallel(cl)
# Create a list to store the results
fis_list <- foreach(i = 1:n_rep, .combine = 'c') %dopar% {
   library(dplyr)
  library(pegas)
  # Calculate the statistics for the i-th matrix
  result_fis_stats <- as_tibble(Fst(as.loci(es[[i]]))) %>% select(Fis)
  return(result_fis_stats$Fis)
}

# Combine the results into a matrix
fis_df <- matrix(unlist(fis_list), nrow = n_rep, byrow = TRUE)
fis_df <- t(fis_df)
# Stop the parallel processing cluster
stopCluster(cl)

# Set row names as in result_f_stats
rownames(fis_df) <- rownames(result_f_stats)
#fis_df <-fis_df[, -1]
vec <- seq(1, n_rep)
colnames(fis_df) <- vec


# Initialize an empty data frame to store the counts
fis_df_Greater <- numeric(sequence_length)
fis_df_Smaller <- numeric(sequence_length)

# Compare the values in result_f_stats[1] to result_FST for each column
for (col in colnames(fis_df)) {
  greater_count <- as_tibble(result_f_stats[2] > fis_df[,col])
  smaller_count <- as_tibble(result_f_stats[2] < fis_df[,col])
  fis_df_Greater <- cbind(fis_df_Greater, greater_count$`Fis (W&C)`)
  fis_df_Smaller <- cbind(fis_df_Smaller, smaller_count$`Fis (W&C)`)
}

# Set row names as in result_f_stats
rownames(fis_df_Smaller) <- rownames(fis_df_Greater) <- rownames(result_f_stats)
fis_df_Smaller <-fis_df_Smaller[, -1]
fis_df_Greater <-fis_df_Greater[, -1]
vec <- seq(1, n_rep)
colnames(fis_df_Smaller) <- colnames(fis_df_Greater) <- vec


fis_df_Smaller_av <- as.data.frame(fis_df_Smaller) %>%
   mutate(average = rowSums(across(where(is.numeric)))/n_rep) %>% select(average)
fis_df_Greater_av <- as.data.frame(fis_df_Greater) %>%
   mutate(average = rowSums(across(where(is.numeric)))/n_rep) %>% select(average)

rownames(fis_df_Smaller_av) <- rownames(fis_df_Greater_av) <- rownames(result_f_stats)

fis_df_sg <- merge(fis_df_Smaller_av, fis_df_Greater_av, by="row.names",all.x=TRUE) %>% column_to_rownames('Row.names')

colnames(fis_df_sg) <- c("smaller", "greater")

fis_df_sg_t <- fis_df_sg %>% mutate(`Two-sided p-values` = ifelse(smaller > 0.5, 2 * (1 - smaller), 2 * smaller))

toc()

# standard deviation SE
# Calculate the standard deviation for each element in fis_df
std_dev <- as.data.frame(apply(fis_df, 1, sd))
colnames(std_dev) <- ("std_dev")

################################# Calculate SE #################################
# join df 
std_dev <- std_dev %>% mutate("SE" = std_dev/sqrt(n_rep)*4)
rownames(std_dev) <- c(rownames(fis_df_sg_t))

df_merge <- merge(std_dev,fis_df_sg_t, by="row.names",all.x=TRUE) %>% column_to_rownames('Row.names')
df_merge <- merge(df_merge, result_f_stats[2],  by="row.names",all.x=TRUE) %>%
  column_to_rownames('Row.names')
df_merge <- df_merge %>% 
  mutate("t" = qt(1-0.05/2, n_pop-1)) %>%
  mutate("95%CI_i" = `Fis (W&C)` - t * SE ) %>% 
  mutate("95%CI_s" = `Fis (W&C)` + t*SE) %>%
  mutate()

```


```{r shuffle_graphm , tidy = TRUE, cache=TRUE}

# plot the point plot 
library(ggplot2)

# Create a ggplot using df_merge
p <- ggplot(df_merge, aes(x = rownames(df_merge), y = `Fis (W&C)`)) +
  geom_point() +
  geom_errorbar(aes(ymin = `95%CI_i`, ymax = `95%CI_s`), width = 0.2, position = position_dodge(0.05)) +
  xlab("Category") +
  ylab("Fis (W&C)")

# Rotate X-axis labels for better readability
p + theme(axis.text.x = element_text(angle = 45, hjust = 1))

```















######################## DEVELOPMENT ######################## 
################# not ready for deployment ##################

#### File export 

```{r export, eval = FALSE}
library(radiator)

genomic_converter(
  data,
  strata = NULL,
  output = NULL,
  filename = NULL,
  parallel.core = parallel::detectCores() - 1,
  verbose = TRUE,
  ...
)

library(hierfstat)
write.fstat
write.struct

```
